### PR TITLE
update gitlab4j and use jersey2 and jackson2 plugins

### PR DESCRIPTION
**Pull request by dcendents**
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

Updated the plugin to use the `jersey2-api` and `jackson2-api` plugins instead of packaging its own version of jersey.

~~This PR requires a modification of `gitlab4j-api` and also `jersey2-api`~~
~~https://github.com/gitlab4j/gitlab4j-api/pull/795~~
~~https://github.com/jenkinsci/jersey2-api-plugin/pull/9~~

fixes #8 

Once this PR is merged and released, we also need to release `gitlab-branch-source-plugin` with the following PR

https://github.com/jenkinsci/gitlab-branch-source-plugin/pull/188

This will fix the following error I had if I installed the `jersey2-api` plugin in jenkins:

```
org.glassfish.jersey.internal.inject.InjectionManagerFactory
java.lang.ClassCastException: Cannot cast org.glassfish.jersey.inject.hk2.Hk2InjectionManagerFactory to org.glassfish.jersey.internal.inject.InjectionManagerFactory
	at java.base/java.lang.Class.cast(Class.java:3605)
	at org.glassfish.jersey.internal.ServiceFinder$LazyObjectIterator.hasNext(ServiceFinder.java:690)
	at org.glassfish.jersey.internal.inject.Injections.lookupService(Injections.java:88)
	at org.glassfish.jersey.internal.inject.Injections.lookupInjectionManagerFactory(Injections.java:73)
	at org.glassfish.jersey.internal.inject.Injections.createInjectionManager(Injections.java:44)
	at org.glassfish.jersey.client.ClientConfig$State.initRuntime(ClientConfig.java:412)
	at org.glassfish.jersey.internal.util.collection.Values$LazyValueImpl.get(Values.java:317)
	at org.glassfish.jersey.client.ClientConfig.getRuntime(ClientConfig.java:807)
	at org.glassfish.jersey.client.ClientRequest.getClientRuntime(ClientRequest.java:219)
	at org.glassfish.jersey.client.ClientRequest.getInjectionManager(ClientRequest.java:610)
	at org.glassfish.jersey.client.JerseyWebTarget.onBuilder(JerseyWebTarget.java:364)
	at org.glassfish.jersey.client.JerseyWebTarget.request(JerseyWebTarget.java:192)
	at org.glassfish.jersey.client.JerseyWebTarget.request(JerseyWebTarget.java:36)
	at org.gitlab4j.api.GitLabApiClient.invocation(GitLabApiClient.java:783)
	at org.gitlab4j.api.GitLabApiClient.invocation(GitLabApiClient.java:748)
	at org.gitlab4j.api.GitLabApiClient.get(GitLabApiClient.java:399)
	at org.gitlab4j.api.GitLabApiClient.get(GitLabApiClient.java:387)
	at org.gitlab4j.api.AbstractApi.get(AbstractApi.java:213)
Caused: org.gitlab4j.api.GitLabApiException: Cannot cast org.glassfish.jersey.inject.hk2.Hk2InjectionManagerFactory to org.glassfish.jersey.internal.inject.InjectionManagerFactory
	at org.gitlab4j.api.AbstractApi.handle(AbstractApi.java:655)
	at org.gitlab4j.api.AbstractApi.get(AbstractApi.java:215)
	at org.gitlab4j.api.RepositoryApi.getBranch(RepositoryApi.java:104)
	at io.jenkins.plugins.gitlabbranchsource.GitLabSCMSource.retrieve(GitLabSCMSource.java:258)
	at jenkins.scm.api.SCMSource.fetch(SCMSource.java:582)
	at io.jenkins.plugins.gitlabbranchsource.helpers.GitLabPipelineStatusNotifier$JobScheduledListener.lambda$onEnterWaiting$0(GitLabPipelineStatusNotifier.java:392)
	at jenkins.util.ContextResettingExecutorService$1.run(ContextResettingExecutorService.java:28)
	at jenkins.security.ImpersonatingExecutorService$1.run(ImpersonatingExecutorService.java:68)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

### CODE REVIEW

**Review by jetersen — APPROVED** [review]

**Comment by jetersen** [comment]
@dcendents you should be able to deploy snapshots to Jenkins's artifactory "repo.jenkins-ci.org"

**Comment by dcendents** [comment]
@jetersen Thanks I didn't know I could:

https://repo.jenkins-ci.org/snapshots/io/jenkins/plugins/jersey2-api/2.35-4-SNAPSHOT/jersey2-api-2.35-4-20220211.131918-1.hpi
https://repo.jenkins-ci.org/snapshots/io/jenkins/plugins/gitlab-api/1.0.7-SNAPSHOT/gitlab-api-1.0.7-20220211.132205-1.hpi

**Comment by jetersen** [comment]
Should be able to use them so Jenkins would pick them up

**Comment by mymarche** [comment]
I installed this versions on my Jenkins, and this works as expected.
@dcendents, Thank you!

**Comment by alexjfisher** [comment]
@dcendents Amazing!  Thanks for looking at this. This should fix https://github.com/jenkinsci/gitlab-branch-source-plugin/issues/151

Meanwhile... a version including https://github.com/gitlab4j/gitlab4j-api/pull/806 is possibly needed much more urgently.  *all* my gitlab builds have started failing in the last 24 hours.  I'm going to be working on setting up my build environment today, to see if I can build my own snapshot plugins.  But should you beat me to it, I sure there are loads of people out there who would be extremely grateful! :)

**Comment by alexjfisher** [comment]
> @alexjfisher here you go: https://repo.jenkins-ci.org/snapshots/io/jenkins/plugins/gitlab-api/1.0.7-SNAPSHOT/gitlab-api-1.0.7-20220302.133039-2.hpi
> 
> This includes gitlab4j-api with both PRs
> 
> * [configure jersey to use jackson over json binding gitlab4j/gitlab4j-api#795](https://github.com/gitlab4j/gitlab4j-api/pull/795)
> * [Fix gitlab4j#805 Increase CommitStatus id from Integer to Long gitlab4j/gitlab4j-api#806](https://github.com/gitlab4j/gitlab4j-api/pull/806)
> 
> Obviously I haven't had time to test it, but `gitlab-branch-source-plugin` does not seem to be affected by this change (no compilation error).

Great!  I'll test this shortly and let you know.  Many thanks.

**Comment by dcendents** [comment]
This just got a bit more complicated, so here are the plugins that need to be installed if you cannot wait for the official release:

- https://repo.jenkins-ci.org/snapshots/io/jenkins/plugins/gitlab-api/1.0.7-SNAPSHOT/gitlab-api-1.0.7-20220302.133039-2.hpi
- https://repo.jenkins-ci.org/snapshots/io/jenkins/plugins/jersey2-api/2.35-5-SNAPSHOT/jersey2-api-2.35-5-20220302.151210-1.hpi
- https://repo.jenkins-ci.org/snapshots/org/jenkins-ci/plugins/jackson2-api/2.13.1-999999-SNAPSHOT/jackson2-api-2.13.1-999999-20220302.151333-1.hpi

Also if you have the `gitlab-plugin` installed, the updates on jersey2-api` and `jackson2-api` will break it, so the following snapshot must be installed as well

- https://repo.jenkins-ci.org/snapshots/org/jenkins-ci/plugins/gitlab-plugin/1.5.29-SNAPSHOT/gitlab-plugin-1.5.29-20220302.151459-1.hpi

**Comment by jetersen** [comment]
@dcendents Thanks for putting it together, looks like a good cleanup :)

**Comment by dcendents** [comment]
> I installed all of the plugins and got the following error (Jenkins 2.336)
> 
> `java.lang.ClassCastException: class org.glassfish.jersey.jaxb.internal.JaxbAutoDiscoverable cannot be cast to class org.glassfish.jersey.internal.spi.AutoDiscoverable (org.glassfish.jersey.jaxb.internal.JaxbAutoDiscoverable ...

Yes I got the same error because the `jira` plugin also has its own version of `jersey`. I deactivated it, we don't really need it. But the fix will be for `jira` to also use the `jersey2-api` plugin instead of packaging its own lib. So you need to hunt down which other plugins have `jersey` libs and deactivate them.

I submitted this PR because if I updated the `gitlab-plugin` from version 1.5.25 onwards it would install the `jersey2-api` plugin and I started having `java.lang.ClassCastException: Cannot cast org.glassfish.jersey.inject.hk2.Hk2InjectionManagerFactory to org.glassfish.jersey.internal.inject.InjectionManagerFactory`

It seems all those plugins were ok when they all had their own private `jersey` libs, but since the introduction of `jersey2-api` no plugin can safely package `jersey` anymore.

I suspect this is because `jersey` seems to rely a lot on `ClassLoader loader = Thread.currentThread().getContextClassLoader();` and jenkins does not like this: 

https://github.com/eclipse-ee4j/jersey/search?p=1&q=getContextClassLoader

https://github.com/eclipse-ee4j/jersey/blob/7b6d2f84391f24310baae7d8e55e644703642826/media/multipart/src/main/java/org/glassfish/jersey/media/multipart/MultiPartProperties.java#L151

https://www.jenkins.io/doc/developer/plugin-development/dependencies-and-class-loading/

**Comment by albertvveld** [comment]
Removed Jira plugin, it is working again thanks!

**Comment by callum-p** [comment]
Sorry to spam up this pull request but I just needed to thank @dcendents. That's resolved my issues that were causing some very inconvenient workarounds this past week.

**Comment by dcendents** [comment]
quick update, my pull requests on `jersey2-api` and `jackson2-api` plugins have been merged and released. However the new version of `gitlab4j-api` modified all the java id fields from `int` to `long`.

So I also need to submit a PR to `gitlab-branch-source-plugin` to work with the updated lib. 

I suspect everyone waiting for this PR also use the `gitlab-branch-source-plugin` (AFAIK it is the only plugin using the `gitlab-api` plugin).

**Comment by dcendents** [comment]
@jetersen This is now ready with the following PR as well:

https://github.com/jenkinsci/gitlab-branch-source-plugin/pull/188

**Comment by jetersen** [comment]
The bom has yet update the jersey-api plugin: https://github.com/jenkinsci/bom/pull/1011

So the PR is not there yet

**Comment by jetersen** [comment]
@dcendents if you want you could pin the version till the bom get sorted out.

File: pom.xml [hunk 1]
```diff
@@ -4,7 +4,7 @@
     <parent>
         <groupId>org.jenkins-ci.plugins</groupId>
         <artifactId>plugin</artifactId>
-        <version>3.47</version>
+        <version>4.33</version>
```
Comment (jetersen) [inline pom.xml:7]:
```suggestion
        <version>4.40</version>
```

File: pom.xml [hunk 2]
```diff
@@ -16,7 +16,7 @@
     <properties>
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
         <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
-        <jenkins.version>2.60.3</jenkins.version>
+        <jenkins.version>2.289.1</jenkins.version>
         <java.level>8</java.level>
```
Comment (jetersen) [inline pom.xml:20]: Removing java.level is required by `4.40`
```suggestion
        <jenkins.version>2.289.3</jenkins.version>
```

I suggest looking at https://github.com/jenkinsci/gitlab-branch-source-plugin/pull/189

File: pom.xml [hunk 3]
```diff
@@ -56,15 +56,60 @@
...
-            <version>4.0.1</version>
+        </dependency>
+        <dependency>
+            <groupId>io.jenkins.plugins</groupId>
+            <artifactId>jersey2-api</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.jenkins-ci.plugins</groupId>
+            <artifactId>jackson2-api</artifactId>
+        </dependency>
```
Comment (jetersen) [inline pom.xml:94]: Temporarily pin these two until BOM can release these.

```suggestion
        <dependency>
            <groupId>io.jenkins.plugins</groupId>
            <artifactId>jersey2-api</artifactId>
            <!-- TODO https://github.com/jenkinsci/bom/pull/1011 -->
            <version>2.35-6</version>
        </dependency>
        <dependency>
            <groupId>org.jenkins-ci.plugins</groupId>
            <artifactId>jackson2-api</artifactId>
            <version>2.13.2.20220328-273.v11d70a_b_a_1a_52</version
        </dependency>
```

**Comment by jetersen** [comment]
Please go ahead and rebase with master thanks to @basil updating the POM

File: pom.xml (hunk 2 above)
Comment (basil) [reply pom.xml:20]: Done in #9 so you just need to rebase.

**Comment by basil** [comment]
And after rebasing you should get an incremental build, which you can then reference in the other PR instead of a manually deployed snapshot.

**Comment by jetersen** [comment]
went ahead and rebased 😅

File: pom.xml [hunk 4]
```diff
@@ -68,7 +68,44 @@
...
+            <dependency>
+            <groupId>io.jenkins.plugins</groupId>
+            <artifactId>jersey2-api</artifactId>
+            <!-- TODO https://github.com/jenkinsci/bom/pull/1011 -->
+            <version>2.35-6</version>
+        </dependency>
+        <dependency>
+            <groupId>org.jenkins-ci.plugins</groupId>
+            <artifactId>jackson2-api</artifactId>
+            <version>2.13.2.20220328-273.v11d70a_b_a_1a_52</version
```
Comment (basil) [inline pom.xml:104]:
```suggestion
            <version>2.13.2.20220328-273.v11d70a_b_a_1a_52</version>
```

Comment (jetersen) [reply pom.xml]: Ya copy pasta error. 😓

File: pom.xml [hunk 5]
```diff
@@ -68,7 +68,44 @@
...
+            <version>2.35-6</version>
+        </dependency>
+        <dependency>
+            <groupId>org.jenkins-ci.plugins</groupId>
+            <artifactId>jackson2-api</artifactId>
+            <version>2.13.2.20220328-273.v11d70a_b_a_1a_52</version>
+        </dependency>
+        <dependency>
+            <groupId>org.jenkins-ci.plugins</groupId>
+            <artifactId>apache-httpcomponents-client-4-api</artifactId>
```
Comment (jetersen) [inline pom.xml:108]: @dcendents any reason this was added?

**Review by basil — COMMENTED** [review]
Took a look at the dependency tree of `org.gitlab4j:gitlab4j-api:5.0.1`:

```
[INFO] +- org.gitlab4j:gitlab4j-api:jar:5.0.1:compile
[INFO] |  +- jakarta.activation:jakarta.activation-api:jar:1.2.2:compile
[INFO] |  +- org.glassfish.jersey.inject:jersey-hk2:jar:2.35:compile
[INFO] |  |  +- org.glassfish.hk2:hk2-locator:jar:2.6.1:compile
[INFO] |  |  |  +- org.glassfish.hk2.external:aopalliance-repackaged:jar:2.6.1:compile
[INFO] |  |  |  +- org.glassfish.hk2:hk2-api:jar:2.6.1:compile
[INFO] |  |  |  \- org.glassfish.hk2:hk2-utils:jar:2.6.1:compile
[INFO] |  |  \- org.javassist:javassist:jar:3.25.0-GA:compile
[INFO] |  +- org.glassfish.jersey.connectors:jersey-apache-connector:jar:2.35:compile
[INFO] |  |  \- org.apache.httpcomponents:httpclient:jar:4.5.13:compile
[INFO] |  +- org.glassfish.jersey.media:jersey-media-multipart:jar:2.35:compile
[INFO] |  |  \- org.jvnet.mimepull:mimepull:jar:1.9.13:compile
[INFO] |  +- org.glassfish.jersey.media:jersey-media-json-jackson:jar:2.35:compile
[INFO] |  |  +- org.glassfish.jersey.ext:jersey-entity-filtering:jar:2.35:compile
[INFO] |  |  +- com.fasterxml.jackson.core:jackson-annotations:jar:2.12.2:compile
[INFO] |  |  +- com.fasterxml.jackson.core:jackson-databind:jar:2.12.2:compile
[INFO] |  |  |  \- com.fasterxml.jackson.core:jackson-core:jar:2.12.2:compile
[INFO] |  |  \- com.fasterxml.jackson.module:jackson-module-jaxb-annotations:jar:2.12.2:compile
[INFO] |  |     \- jakarta.xml.bind:jakarta.xml.bind-api:jar:2.3.2:compile
[INFO] |  +- jakarta.servlet:jakarta.servlet-api:jar:4.0.4:compile
[INFO] |  \- uk.org.webcompere:system-stubs-jupiter:jar:1.2.0:compile
[INFO] |     \- uk.org.webcompere:system-stubs-core:jar:1.2.0:compile
```


`jakarta.activation:jakarta.activation-api` should not be bundled in this plugin; rather, exclude it and depend on

```
    <dependency>
      <groupId>io.jenkins.plugins</groupId>
      <artifactId>javax-activation-api</artifactId>
      <version>1.2.0-2</version>
    </dependency>
```

`uk.org.webcompere:system-stubs-jupiter` and `uk.org.webcompere:system-stubs-core` look like mocking/test libraries. Probably an upstream issue that these are not in `<scope>test</scope>`, but in the meantime we should work around that by not bundling them in this Jenkins plugin.

`jakarta.servlet:jakarta.servlet-api` is provided by core and should be excluded.

Please also add comments for each exclusion explaining where the classes come from (core, a plugin, etc).

File: pom.xml (hunk 5 above)
Comment (basil) [reply pom.xml:108]: I figured this out; it's because it's needed by `jersey-apache-connector` but `jersey2-api` declares it as optional.

**Comment by jetersen** [comment]
@basil thanks for digging into it, already created: https://github.com/gitlab4j/gitlab4j-api/pull/840

File: pom.xml [hunk 6]
```diff
@@ -68,7 +68,44 @@
...
         </dependency>
```
Comment (basil) [inline pom.xml:109]:
```suggestion
        <dependency>
            <groupId>io.jenkins.plugins</groupId>
            <artifactId>javax-activation-api</artifactId>
        </dependency>
        <dependency>
            <groupId>io.jenkins.plugins</groupId>
            <artifactId>jersey2-api</artifactId>
            <version>2.35-6</version> <!-- TODO https://github.com/jenkinsci/bom/pull/1011 -->
        </dependency>
        <dependency>
            <groupId>org.gitlab4j</groupId>
            <artifactId>gitlab4j-api</artifactId>
            <version>5.0.1</version>
            <exclusions>
                <!-- Provided by javax-activation-api plugin -->
                <exclusion>
                    <groupId>jakarta.activation</groupId>
                    <artifactId>jakarta.activation-api</artifactId>
                </exclusion>
                <!-- Provided by core -->
                <exclusion>
                    <groupId>jakarta.servlet</groupId>
                    <artifactId>jakarta.servlet-api</artifactId>
                </exclusion>
                <!-- Provided by jersey2-api plugin -->
                <exclusion>
                    <groupId>org.glassfish.jersey.connectors</groupId>
                    <artifactId>jersey-apache-connector</artifactId>
                </exclusion>
                <exclusion>
                    <groupId>org.glassfish.jersey.core</groupId>
                    <artifactId>jersey-client</artifactId>
                </exclusion>
                <exclusion>
                    <groupId>org.glassfish.jersey.inject</groupId>
                    <artifactId>jersey-hk2</artifactId>
                </exclusion>
                <exclusion>
                    <groupId>org.glassfish.jersey.media</groupId>
                    <artifactId>jersey-media-json-jackson</artifactId>
                </exclusion>
                <exclusion>
                    <groupId>org.glassfish.jersey.media</groupId>
                    <artifactId>jersey-media-multipart</artifactId>
                </exclusion>
                <!-- Unneeded test library -->
                <exclusion>
                    <groupId>uk.org.webcompere</groupId>
                    <artifactId>system-stubs-jupiter</artifactId>
                </exclusion>
            </exclusions>
        </dependency>
        <!-- Needed by jersey-apache-connector, but jersey2-api plugin declares it as optional -->
        <dependency>
            <groupId>org.jenkins-ci.plugins</groupId>
            <artifactId>apache-httpcomponents-client-4-api</artifactId>
        </dependency>
        <dependency>
            <groupId>org.jenkins-ci.plugins</groupId>
            <artifactId>jackson2-api</artifactId>
            <version>2.13.2.20220328-273.v11d70a_b_a_1a_52</version> <!-- TODO https://github.com/jenkinsci/bom/pull/1013 -->
        </dependency>
```

Comment (jetersen) [reply pom.xml:109]: `Applying suggestions on deleted lines is currently not supported.` 😭

File: pom.xml [hunk 7]
```diff
@@ -65,10 +65,67 @@
...
+                    <groupId>uk.org.webcompere</groupId>
+                    <artifactId>system-stubs-jupiter</artifactId>
+                </exclusion>
+            </exclusions>
+        </dependency>
+        <!-- Needed by jersey-apache-connector, but jersey2-api plugin declares it as optional -->
+        <dependency>
+            <groupId>org.jenkins-ci.plugins</groupId>
+            <artifactId>apache-httpcomponents-client-4-api</artifactId>
+        </dependency>
```
Comment (jetersen) [inline pom.xml:124]: Should this be fixed in jersey2-api and make it non optional? again `apache-httpcomponents-client-4-api` is pretty in all Jenkins installs anyhow.

Comment (basil) [reply pom.xml:124]: We could. The downside is that someone who wants to use some other parts of Jersey 2 but doesn't need the Jersey Apache connector would be forced to install it. That might not be so bad, after all. I could go either way, don't feel strongly.

Comment (jetersen) [reply pom.xml:124]: Me neither, I was simply wondering so 🤷‍♂️

**Comment by basil** [comment]
Check this out, I ported `gitlab-oauth-plugin` to use this new library in https://github.com/jenkinsci/gitlab-oauth-plugin/pull/66

If feasible (I have no way to test) that would be a huge boon for the maintainability of `gitlab-oauth-plugin`.

**Comment by jetersen** [comment]
@callum-p @albertvveld @mymarche @alexjfisher You should be able to pickup the HPI for gitlab-api and gitlab-branch-source plugin to verify
https://repo.jenkins-ci.org/incrementals/io/jenkins/plugins/gitlab-api/1.0.7-rc71.64a_4c6877d17/
https://repo.jenkins-ci.org/incrementals/io/jenkins/plugins/gitlab-branch-source/1.5.10-rc624.6fb_d0fcf2a_0c/

**Comment by basil** [comment]
Yes, and I know I'm hijacking this thread but if I could have any volunteers to test the incremental from https://github.com/jenkinsci/gitlab-oauth-plugin/pull/66 that would be super cool =)

**Review by basil — APPROVED** [review]

**Comment by jetersen** [comment]
@dcendents 
You might want to add your email on your GitHub account 🥇
https://help.github.com/articles/adding-an-email-address-to-your-github-account/

link to see current commit author: https://github.com/jenkinsci/gitlab-api-plugin/commit/a3711d5d3b1c6b8771b54ca247c8fa7b32452c5c.patch

**Comment by dcendents** [comment]
@basil @jetersen Sorry guys I was not available this weekend but I can see you did all the work.
Thanks a lot

**Comment by jetersen** [comment]
@dcendents your welcome 😅

FYI I have created https://github.com/jenkinsci/jira-plugin/pull/435

**Pull request merged**